### PR TITLE
BUG: GroupBy.apply() throws erroneous ValueError with duplicate axes

### LINKED ISSUE #16646: "cannot reindex from a duplicate axis" when groupby().apply() on MultiIndex columns
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

df = pd.DataFrame(
    np.ones([6, 4], dtype=int),
    columns=pd.MultiIndex.from_product([['A', 'B'], [1, 2]])
)

(
    df
    .groupby(level=0, axis=1)
    .apply(
        lambda df: pd.concat(
            [df.xs(df.name, axis=1), df.sum(axis=1).to_frame('Total')], 
            axis=1
        )
    )
)

```
#### Problem description

The code above produces the error:

> cannot reindex from a duplicate axis

I believe this is a bug because, as described below, essentially the same operations can be successfully run along the other axis.

#### Expected Output

It should, as far as I can tell, produce the following output:

image

The desired output can be obtained when working on the transposed DataFrame along the `index` rather than the `columns`:

```python
(
    df
    .T
    .groupby(level=0)
    .apply(
        lambda df: pd.concat(
            [df.xs(df.name), df.sum().to_frame('Total').T]
        )
    )
    .T
)
```

#### Output of ``pd.show_versions()``

<details>

pd.show_versions()
pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.9.final.0
python-bits: 64
OS: Windows
OS-release: 8
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.2
pytest: None
pip: 9.0.1
setuptools: 34.3.3
Cython: None
numpy: 1.13.0
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.0
openpyxl: None
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999
sqlalchemy: None
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.9.5
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by smithto1**
- [x] closes #16646 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

*Behavioural Change*
`GroupBy.apply` would sometimes throw an erroneous `ValueError: cannot reindex from duplicate axis` because when it checked if the output was mutated from its original shape it would only check the index, when sometimes it should check the columns. 

*Tests*
One new test added on `test_apply.py` which fails on master because of the bug; passes with this PR. One existing test was marked *xfail* but the *xfail* is now removed. One existing test previously had to manipulate the expected index, but that is no longer necessary. 

All of the copy-pasteable examples from #16646 are fixed with this PR.

### CODE REVIEW

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -132,6 +132,7 @@ Plotting
 Groupby/resample/rolling
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
+- Bug in :meth:`DataFrameGroupBy.apply` that would some times throw an erroneous ``ValueError: cannot reindex from a duplicate axis`` (:issue:`16646`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.2.0.rst:135]: Can you add some context as to when this might happen? I think axis=1 and MultiIndex columns, right?

**Review by rhshadrach — COMMENTED** [review]
Thanks for the PR! Looks good - just some minor comments.

File: pandas/tests/groupby/test_apply.py [hunk 2]
```diff
@@ -63,15 +63,9 @@ def test_apply_trivial():
...
-    reason="GH#20066; function passed into apply "
-    "returns a DataFrame with the same index "
-    "as the one to create GroupBy object."
-)
 def test_apply_trivial_fail():
-    # GH 20066
-    # trivial apply fails if the constant dataframe has the same index
-    # with the one used to create GroupBy object.
+    # GH 35441
+    # test passes, xfail removed
```
[the remark this answers is not in the document]
Comment (smithto1) [reply pandas/tests/groupby/test_apply.py]: Done.

File: doc/source/whatsnew/v1.2.0.rst (hunk 1 above)
Comment (smithto1) [reply doc/source/whatsnew/v1.2.0.rst]: Done.

File: pandas/tests/groupby/test_apply.py [hunk 3]
```diff
@@ -63,15 +63,8 @@ def test_apply_trivial():
...
-@pytest.mark.xfail(
-    reason="GH#20066; function passed into apply "
-    "returns a DataFrame with the same index "
-    "as the one to create GroupBy object."
-)
 def test_apply_trivial_fail():
-    # GH 20066
-    # trivial apply fails if the constant dataframe has the same index
-    # with the one used to create GroupBy object.
+    # GH 35441
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_apply.py:67]: I think this should be left as 20066. That issue explains why the test is here, not this PR.

**Review by rhshadrach — COMMENTED** [review]
lgtm, one small detail I missed previously.

File: pandas/tests/groupby/test_apply.py (hunk 3 above)
Comment (smithto1) [reply pandas/tests/groupby/test_apply.py]: Done.

**Comment by jreback** [comment]
lgtm. pls merge master and ping on green.

**Comment by smithto1** [comment]
@jreback I think this is ready to merge.

**Comment by jreback** [comment]
thanks @smithto1 very nice

File: pandas/tests/groupby/test_function.py [hunk 4]
```diff
@@ -940,10 +940,6 @@ def test_frame_describe_multikey(tsframe):
     groupedT = tsframe.groupby({"A": 0, "B": 0, "C": 1, "D": 1}, axis=1)
     result = groupedT.describe()
     expected = tsframe.describe().T
-    expected.index = pd.MultiIndex(
```
Comment (TomAugspurger) [inline pandas/tests/groupby/test_function.py:943]: > One existing test previously had to manipulate the expected index, but that is no longer necessary.

@smithto1 I don't understand the change to the expected here, can you expand on why the new value is correct?

Comment (smithto1) [reply pandas/tests/groupby/test_function.py:943]: With the bugfix in this PR, I think `.describe()` on `axis=1` is now encountering the bug reported in #26805, #22848 #22546 where `group_keys=True` is ignored. 

The behaviour of the new code with `group_keys=True/False` is the same as the old code with `group_keys=False`.

I'll try to take a look at the 'group_keys' bug; if that is fixed I think this test will probably be reverted.

Comment (smithto1) [reply pandas/tests/groupby/test_function.py:943]: I think this may be addressed by #34998 but for the `axis=1` case.

**Pull request merged**